File: frontend/src/api/tax.ts
import { apiGet, apiPost } from "./client";

// One do-as-we-go item tax classification row, matching the backend's
// admin tax classifications endpoint (api/tax_classifications.py --
// see docs/design/16-sales-tax.md). `status` starts "pending" whenever
// the classifier engine has never seen this normalized item key before;
// an admin either confirms the model's guess as-is or overrides it with
// a corrected category/taxable/hts_code, both of which move it out of
// the pending review queue.
export interface TaxClassification {
  id: string;
  normalized_key: string;
  description: string;
  category: string;
  taxable: boolean;
  hts_code: string | null;
  status: "pending" | "confirmed" | "overridden";
  source: string;
  model: string | null;
  rationale: string | null;
  confirmed_at: string | null;
  updated_at: string;
}

// Omit `status` to fetch every classification regardless of review state.
export function listTaxClassifications(
  status?: "pending" | "confirmed" | "overridden",
): Promise<TaxClassification[]> {
  const params = status ? `?${new URLSearchParams({ status }).toString()}` : "";
  return apiGet<TaxClassification[]>(`/api/admin/tax/classifications${params}`);
}

// normalized_key may contain spaces/punctuation -- always URL-encoded here
// so callers never have to remember to do it themselves.
export function confirmTaxClassification(
  normalizedKey: string,
): Promise<TaxClassification> {
  return apiPost<TaxClassification>(
    `/api/admin/tax/classifications/${encodeURIComponent(normalizedKey)}/confirm`,
  );
}

export interface TaxClassificationOverrideInput {
  category: string;
  taxable: boolean;
  hts_code?: string | null;
}

export function overrideTaxClassification(
  normalizedKey: string,
  body: TaxClassificationOverrideInput,
): Promise<TaxClassification> {
  return apiPost<TaxClassification>(
    `/api/admin/tax/classifications/${encodeURIComponent(normalizedKey)}/override`,
    body,
  );
}

// Stripe's own recorded tax figures for the same range as the deterministic
// tax report -- a cross-check, not a source of truth; only covers
// Stripe-processed payments (see api/invoices.py's stripe-reconcile route).
export interface StripeTaxReconcile {
  total_tax_collected: string;
  by_jurisdiction: Record<string, string>;
  transaction_count: number;
}

export function getStripeReconcile(
  fromDate: string,
  toDate: string,
): Promise<StripeTaxReconcile> {
  const params = new URLSearchParams({ from_date: fromDate, to_date: toDate });
  return apiGet<StripeTaxReconcile>(
    `/api/admin/tax/stripe-reconcile?${params.toString()}`,
  );
}

// One rate row in the tax_rules knowledge base (docs/design/16-sales-tax.md)
// -- admin-entered and government-cited. Claude only ever classifies items
// into a category; it never sets or approves the rate itself, so every row
// here traces back to a real citation_url an admin typed in.
export interface TaxRule {
  id: string;
  jurisdiction: string;
  tax_type: string;
  category: string;
  rate: string;
  source: string;
  citation_url: string | null;
  effective_from: string;
}

export function listTaxRules(): Promise<TaxRule[]> {
  return apiGet<TaxRule[]>("/api/admin/tax/rules");
}

export interface TaxRuleInput {
  jurisdiction: string;
  tax_type: string;
  category?: string;
  rate: string;
  source: string;
  citation_url: string;
}

export function addTaxRule(input: TaxRuleInput): Promise<TaxRule> {
  return apiPost<TaxRule>("/api/admin/tax/rules", input);
}

File: android/core/src/main/kotlin/app/logand/core/model/Tax.kt
package app.logand.core.model

import kotlinx.serialization.Serializable

// Every model in this file is copy-derived from a real backend admin route
// response/request shape (never guessed) -- see backend's api/tax.py
// (prefix /api/admin/tax) and api/invoices.py's tax-report route (on the
// invoices router, not the tax one). Money/rate fields travel as strings
// (Decimal -> JSON string), same convention as the rest of Admin.kt.

// -- item tax classifications ---------------------------------------------

// One do-as-we-go item tax classification row, matching
// api/tax.py::_serialize. `status` starts "pending" whenever the
// classifier engine has never seen this normalized item key before; an
// admin either confirms the model's guess as-is or overrides it with a
// corrected category/taxable/hts_code, both of which move it out of the
// pending review queue. `status` is left a plain String (not a Kotlin
// enum) so an unrecognized-by-this-client-version value from a future
// backend never fails to decode.
@Serializable
data class TaxClassification(
    val id: String,
    val normalized_key: String,
    val description: String,
    val category: String,
    val taxable: Boolean,
    val hts_code: String?,
    val status: String,
    val source: String,
    val model: String?,
    val rationale: String?,
    val confirmed_at: String?,
    val updated_at: String?,
)

// Body for POST /api/admin/tax/classifications/{key}/override -- matches
// api/tax.py::OverrideInput. Confirming/overriding a tax classification
// changes financial records, so callers must only send this after an
// explicit confirm step in the UI, never a one-tap action.
@Serializable
data class TaxClassificationOverrideRequest(
    val category: String,
    val taxable: Boolean,
    val hts_code: String? = null,
)

// -- Stripe tax reconciliation ---------------------------------------------

// Response of GET /api/admin/tax/stripe-reconcile -- Stripe's own recorded
// tax figures for a date range, for cross-checking against the
// deterministic tax report's own figures for the same period. Best-effort
// on the backend: an unconfigured Stripe account or any failure in the
// round trip returns zeros, never a 5xx (see stripe_reconcile.py).
@Serializable
data class StripeTaxReconcile(
    val total_tax_collected: String,
    val by_jurisdiction: Map<String, String>,
    val transaction_count: Int,
)

// -- tax rules knowledge base -----------------------------------------------

// One rate row in the tax_rules knowledge base, matching
// api/tax.py::_serialize_rule -- admin-entered and government-cited.
// Claude only ever classifies items into a category; it never sets or
// approves the rate itself, so every row here traces back to a real
// citation_url an admin typed in.
@Serializable
data class TaxRule(
    val id: String,
    val jurisdiction: String,
    val tax_type: String,
    val category: String,
    val rate: String,
    val source: String,
    val citation_url: String?,
    val effective_from: String,
)

// Body for POST /api/admin/tax/rules -- matches
// api/tax.py::TaxRuleCreateInput. citation_url is required by the
// backend (rejects anything else with a 400 -- see migration
// 0027_tax_rule_citation); the UI must validate this is non-blank
// before ever making the call, not rely on the network round trip to
// surface it.
@Serializable
data class TaxRuleCreateRequest(
    val jurisdiction: String,
    val tax_type: String,
    val category: String = "*",
    val rate: String,
    val source: String,
    val citation_url: String,
)

// -- tax filing report -------------------------------------------------------

// Matches domain/invoices/tax/report.py::TaxReport's jurisdiction rows, as
// serialized by api/invoices.py::get_tax_report.
@Serializable
data class TaxReportJurisdictionRow(
    val jurisdiction: String,
    val tax_type: String,
    val taxable_base: String,
    val tax_collected: String,
)

@Serializable
data class TaxReportCategoryRow(
    val category: String,
    val gross: String,
    val taxable_gross: String,
)

// Response of GET /api/admin/invoices/tax-report -- tax-filing breakdown
// over [from_date, to_date): sales by category, tax collected by
// jurisdiction + type, and which jurisdictions must be filed for. See
// docs/design/16-sales-tax.md.
@Serializable
data class TaxReport(
    val from_date: String,
    val to_date: String,
    val currency: String,
    val invoice_count: Int,
    val total_sales: String,
    val total_tax_collected: String,
    val filing_jurisdictions: List<String>,
    val by_jurisdiction: List<TaxReportJurisdictionRow>,
    val by_category: List<TaxReportCategoryRow>,
)
